File: Windows/messagemanager.cpp
/*
 * Title:  spider messagemanager.cpp (Windows)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "messagemanager.hpp"
#include "clientmanager.hpp"
#include "servermanager.hpp"
#include "pipemanager.hpp"
#include "routingmanager.hpp"
#include "socks5message.hpp"
#include "message.hpp"
#include "routingmessagequeue.hpp"
#include "socks5messagequeue.hpp"
#include "routingmessage.hpp"
#include "client.hpp"
#include "server.hpp"
#include "pipe.hpp"


namespace spider
{
    Messagemanager::Messagemanager(std::string spider_ip,
                                   std::shared_ptr<Clientmanager> client_manager,
                                   std::shared_ptr<Servermanager> server_manager,
                                   std::shared_ptr<Pipemanager> pipe_manager,
                                   std::shared_ptr<Routingmanager> routing_manager)
    {
        this->spider_ip = spider_ip;
        this->client_manager = client_manager;
        this->server_manager = server_manager;
        this->pipe_manager = pipe_manager;
        this->routing_manager = routing_manager;
        this->routing_messages_queue = std::make_unique<Routingmessagequeue>();
        this->socks5_messages_queue = std::make_unique<Socks5messagequeue>();
    }

    Messagemanager::~Messagemanager()
    {

    }

    void Messagemanager::push_routing_message(std::shared_ptr<Routingmessage> routing_message)
    {
        routing_messages_queue->push(routing_message);

        return;
    }

    int32_t Messagemanager::push_timeout_routing_message(std::shared_ptr<Routingmessage> routing_message,
                                                         int32_t tv_sec,
                                                         int32_t tv_usec)
    {
        int32_t ret = 0;

        ret = routing_messages_queue->push_timeout(routing_message,
                                                   tv_sec,
                                                   tv_usec);

        return ret;
    }

    void Messagemanager::push_socks5_message(std::shared_ptr<Socks5message> socks5_message)
    {
        socks5_messages_queue->push(socks5_message);

        return;
    }

    int32_t Messagemanager::push_timeout_socks5_message(std::shared_ptr<Socks5message> socks5_message,
                                                        int32_t tv_sec,
                                                        int32_t tv_usec)
    {
        int32_t ret = 0;

        ret = socks5_messages_queue->push_timeout(socks5_message,
                                                  tv_sec,
                                                  tv_usec);

        return ret;
    }

    std::shared_ptr<Routingmessage> Messagemanager::pop_routing_message()
    {
        std::shared_ptr<Routingmessage> routing_message = routing_messages_queue->pop();

        return routing_message;
    }

    std::shared_ptr<Socks5message> Messagemanager::pop_socks5_message()
    {
        std::shared_ptr<Socks5message> socks5_message = socks5_messages_queue->pop();

        return socks5_message;
    }

    void Messagemanager::transfer_routing_message()
    {
        char message_type;
        std::shared_ptr<Routingmessage> routing_message;


        routing_message = pop_routing_message();
        if(routing_message != nullptr )
        {
            message_type = routing_message->get_message_type();
            if(message_type == 'r')  // routing message
            {
                routing_manager->push_routing_message(routing_message);
            }else
            {
#ifdef DEBUGPRINT
                std::printf("[-] unknown message type: %c\n",
                            message_type);
#endif
            }
        }

        return;
    }

    std::shared_ptr<Socks5message> Messagemanager::transfer_socks5_message()
    {
        int ret = 0;
        char message_type;
        std::shared_ptr<Socks5message> socks5_message;
        char destination_node_type;
        std::shared_ptr<Client> client;
        std::shared_ptr<Server> server;
        std::shared_ptr<Pipe> pipe;
        uint32_t connection_id = 0;
        uint32_t client_id = 0;
        uint32_t server_id = 0;
        uint32_t pipe_id = 0;


        socks5_message = pop_socks5_message();
        if(socks5_message != nullptr)
        {
            message_type = socks5_message->get_message_type();
            if(message_type == 's')  // socks5 message
            {
                if(spider_ip == socks5_message->get_destination_ip())
                {
                    connection_id = socks5_message->get_connection_id();
                    client_id = socks5_message->get_client_id();
                    server_id = socks5_message->get_server_id();
                    destination_node_type = socks5_message->get_destination_node_type();
                    if(destination_node_type == 'c')    // client
                    {
                        client = client_manager->get_client(connection_id,
                                                            client_id);
                        if(client != nullptr)
                        {
//                            client->push_socks5_message(socks5_message);
                            std::thread thread_client(&Client::push_socks5_message,
                                                      client,
                                                      socks5_message);
                            thread_client.detach();

                        }else
                        {
#ifdef DEBUGPRINT
                            std::printf("[-] cannot transfer client message\n");
#endif
                        }
                    }else if(destination_node_type == 's')  // server
                    {
                       server = server_manager->get_server(connection_id,
                                                            server_id);
                        if(server != nullptr)
                        {
//                            server->push_socks5_message(socks5_message);
                            std::thread thread_server(&Server::push_socks5_message,
                                                      server,
                                                      socks5_message);
                            thread_server.detach();
                        }else
                        {
                            // generate server
                            return socks5_message;
                        }
                    }else
                    {
#ifdef DEBUGPRINT
                        std::printf("[-] cannot transfer socks5 message\n");
#endif
                    }
                }else
                {
                    pipe = routing_manager->get_destination_pipe(socks5_message->get_destination_node_type(),
                                                                 socks5_message->get_destination_ip());
                    if(pipe != nullptr)
                    {
//                        pipe->push_socks5_message(socks5_message);
                        std::thread thread_pipe(&Pipe::push_socks5_message,
                                                pipe,
                                                socks5_message);
                        thread_pipe.detach();
                    }else
                    {
#ifdef DEBUGPRINT
                        std::printf("[-] cannot transfer pipe message\n");
#endif
                    }
                }
            }else
            {
#ifdef DEBUGPRINT
                std::printf("[-] unknown message type: %c\n",
                            message_type);
#endif
            }
        }

        return nullptr;
    }
}

File: Windows_csharp/MessageManager.cs
/*
 * Title:  spider MessageManager.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Threading;

namespace spider
{
    public class MessageManager
    {
        private SpiderIp spiderIp;
        private ClientManager clientManager;
        private ServerManager serverManager;
        private PipeManager pipeManager;
        private RoutingManager routingManager;
        private RoutingMessageQueue routingMessagesQueue;
        private Socks5MessageQueue socks5MessagesQueue;

        public MessageManager(SpiderIp spiderIp,
                              ClientManager clientManager,
                              ServerManager serverManager,
                              PipeManager pipeManager,
                              RoutingManager routingManager)
        {
            this.spiderIp = spiderIp;
            this.clientManager = clientManager;
            this.serverManager = serverManager;
            this.pipeManager = pipeManager;
            this.routingManager = routingManager;
            this.routingMessagesQueue = new RoutingMessageQueue();
            this.socks5MessagesQueue = new Socks5MessageQueue();
        }

        ~MessageManager()
        {

        }

        public SpiderIp SpiderIp
        {
            get { return spiderIp; }
            set { spiderIp = value; }
        }

        public void PushRoutingMessage(RoutingMessage routingMessage)
        {
            routingMessagesQueue.Push(routingMessage);

            return;
        }

        public void PushRoutingMessage(Object obj)
        {
            RoutingMessage routingMessage = obj as RoutingMessage;


            if(routingMessage != null)
            {
                routingMessagesQueue.Push(routingMessage);
            }

            return;
        }

        public int PushTimeoutRoutingMessage(RoutingMessage routingMessage,
                                             int tvSec,
                                             int tvUsec)
        {
            int ret = 0;

            ret = routingMessagesQueue.PushTimeout(routingMessage,
                                                   tvSec,
                                                   tvUsec);

            return ret;
        }

        public void PushSocks5Message(Socks5Message socks5Message)
        {
            socks5MessagesQueue.Push(socks5Message);

            return;
        }

        public void PushSocks5Message(Object obj)
        {
            Socks5Message socks5Message = obj as Socks5Message;


            if(socks5Message != null)
            {
                socks5MessagesQueue.Push(socks5Message);
            }

            return;
        }

        public int PushTimeoutSocks5Message(Socks5Message socks5Message,
                                            int tvSec,
                                            int tvUsec)
        {
            int ret = 0;

            ret = socks5MessagesQueue.PushTimeout(socks5Message,
                                                  tvSec,
                                                  tvUsec);

            return ret;
        }

        private RoutingMessage PopRoutingMessage()
        {
            RoutingMessage routingMessage = routingMessagesQueue.Pop();

            return routingMessage;
        }

        private Socks5Message PopSocks5Message()
        {
            Socks5Message socks5Message = socks5MessagesQueue.Pop();

            return socks5Message;
        }

        public void TransferRoutingMessage()
        {
            char messageType;
            RoutingMessage routingMessage;

            routingMessage = PopRoutingMessage();
            if(routingMessage != null)
            {
                messageType = routingMessage.MessageType;
                if(messageType == 'r')  // routing message
                {
                    routingManager.PushRoutingMessage(routingMessage);
                }else
                {
#if DEBUGPRINT
                    Console.WriteLine("[-] unknown message type: {0}",
                                      messageType);
#endif
                }
            }

            return;
        }

        public Socks5Message TransferSocks5Message()
        {
            char messageType;
            Socks5Message socks5Message;
            char destinationNodeType;
            Client client;
            Server server;
            Pipe pipe;
            uint connectionId = 0;
            uint clientId = 0;
            uint serverId = 0;


            socks5Message = PopSocks5Message();
            if(socks5Message != null)
            {
                messageType = socks5Message.MessageType;
                if(messageType == 's')
                {
                    if((String.Compare(spiderIp.SpiderIpv4, socks5Message.DestinationIp) == 0) ||
                       (String.Compare(spiderIp.SpiderIpv6Global, socks5Message.DestinationIp) == 0) ||
                       (String.Compare(spiderIp.SpiderIpv6UniqueLocal, socks5Message.DestinationIp) == 0) ||
                       (String.Compare(spiderIp.SpiderIpv6LinkLocal, socks5Message.DestinationIp) == 0))
                    {
                        connectionId = socks5Message.ConnectionId;
                        clientId = socks5Message.ClientId;
                        serverId = socks5Message.ServerId;
                        destinationNodeType = socks5Message.DestinationNodeType;
                        if(destinationNodeType == 'c')  // client
                        {
                            client = clientManager.GetClient(connectionId,
                                                             clientId);
                            if(client != null)
                            {
                                Thread threadClient = new Thread(new ParameterizedThreadStart(client.PushSocks5Message));
                                threadClient.Start(socks5Message);
                            }else
                            {
#if DEBUGPRINT
                                Console.WriteLine("[-] cannot transfer client message");
#endif
                            }
                        }else if(destinationNodeType == 's')    // server
                        {
                            server = serverManager.GetServer(connectionId,
                                                             serverId);
                            if(server != null)
                            {
                                Thread threadServer = new Thread(new ParameterizedThreadStart(server.PushSocks5Message));
                                threadServer.Start(socks5Message);
                            }else
                            {
                                // generate server
                                return socks5Message;
                            }
                        }else
                        {
#if DEBUGPRINT
                            Console.WriteLine("[-] cannot transfer socks5 message");
#endif
                        }
                    }else
                    {
                        pipe = routingManager.GetDestinationPipe(socks5Message.DestinationIp);
                        if(pipe != null)
                        {
                            Thread threadPipe = new Thread(new ParameterizedThreadStart(pipe.PushSocks5Message));
                            threadPipe.Start(socks5Message);
                        }else
                        {
#if DEBUGPRINT
                            Console.WriteLine("[-] cannot transfer pipe message");
#endif
                        }
                    }
                }else
                {
#if DEBUGPRINT
                    Console.WriteLine("[-] unknown message type: {0}",
                                      messageType);
#endif
                }
            }

            return null;
        }
    }
}

